File: Go/hoshimeguri2.go
//  ビルド:
//    $ go build hoshimeguri2.go
//
//  実行:
//    $ ./hoshimeguri2 ../data.txt

package main

import (
  "bufio"
  "bytes"
  "fmt"
  "os"
  "strings"
)

const points = "AHCJEBGDIF"

type Star struct {
  indices []int
}

func NewStar(c byte) Star {
  return Star { indices: []int{strings.IndexByte(points, c)} }
}

func (star *Star) Step(c rune) Star {
  index := star.indices[len(star.indices) - 1]
  switch struct{int; rune}{index % 2, c} {
    case struct {int; rune}{1, 'R'}: index = (index + 2) % 10
    case struct {int; rune}{0, 'W'}: index = (index + 8) % 10
    default:                         index = (index + 1) % 10
  }
  star.indices = append(star.indices, index)
  return *star
}

func (star *Star) String() string {
  result := bytes.NewBufferString("")
  for _, index := range star.indices {
    result.WriteByte(points[index])
  }
  return result.String()
}

func solve(input string) string {
  star := NewStar(input[0])
  for _, c := range input[1:] {
    star.Step(c)
  }
  return star.String()
}

func test(input string, expected string) {
  actual := solve(input)
  if actual == expected {
    fmt.Print(".")
  } else {
    fmt.Printf(`
input:    %s
expected: %s
actual:   %s
`, input, expected, actual)
  }
}

func main() {
  filename := os.Args[1]
  file, _ := os.Open(filename)
  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    tokens := strings.Split(scanner.Text(), " ")
    test(tokens[0], tokens[1])
  }
  file.Close()
  fmt.Println()
}

File: Erlang/hoshimeguri2.erl
% ビルド:
%   $ erlc hoshimeguri2.erl
%
% 実行:
%   $ erl -run hoshimeguri2 main --noshell -data ../data.txt

-module(hoshimeguri2).
-export([main/0]).

index(C) -> string:str("AHCJEBGDIF", [C]) - 1.

step(P0, $R) when P0 rem 2 =:= 1 -> (P0 + 2) rem 10;
step(P0, $W) when P0 rem 2 =:= 0 -> (P0 + 8) rem 10;
step(P0, _ ) -> (P0 + 1) rem 10.

solve([FirstPoint|Colors]) ->
  Indices = lists:foldl(
    fun(Color, Indices) ->
      lists:append(Indices, [step(lists:last(Indices), Color)])
    end,
    [index(FirstPoint)],
    Colors
  ),
  lists:map(fun(I) -> lists:nth(I + 1, "AHCJEBGDIF") end, Indices).

test(Input, Expected) ->
  Actual = solve(Input),
  if
    Actual == Expected -> io:put_chars(".");
    true               -> io:format("""

input:    ~s
expected: ~s
actual:   ~s
""", [Input, Expected, Actual])
  end.

test(Line) ->
  [Input, Expected] = string:tokens(Line, [$ , $\n]),
  test(Input, Expected).

read_lines(Handle) ->
  case file:read_line(Handle) of
    {ok, Line} -> [Line|read_lines(Handle)];
    _Otherwise -> []
  end.

main() ->
  {ok, [[Filename]]} = init:get_argument(data),
  {ok, Handle} = file:open(Filename, read),
  Lines = read_lines(Handle),
  file:close(Handle),
  lists:foreach(fun test/1, Lines),
  io:nl(),
  halt().
